File: backend/app/schemas/schema_mapping.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class SemanticRoleDefinition(BaseModel):
    role: str
    label: str
    description: str


class SchemaRoleListResponse(BaseModel):
    items: list[SemanticRoleDefinition]


class ColumnEvidence(BaseModel):
    evidence_type: str
    description: str
    score_contribution: float
    observed_value: str | float | int | bool | None
    expected_pattern: str
    severity: str


class ColumnCandidate(BaseModel):
    role: str
    confidence_score: float
    summary_evidence: list[str]


class SchemaValidationIssue(BaseModel):
    code: str
    message: str
    severity: str
    column_names: list[str] = []


class ColumnProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    column_index: int
    column_name: str
    normalized_column_name: str
    physical_type: str
    semantic_role: str
    confidence_score: float
    confirmation_status: str
    decision_source: str
    nullable: bool
    null_count: int
    sample_count: int
    unique_count: int
    parse_success_rate: float
    numeric_min: float | None
    numeric_max: float | None
    numeric_mean: float | None
    date_min: str | None
    date_max: str | None
    string_min_length: int | None
    string_max_length: int | None
    sample_values: list[str]
    evidence: list[ColumnEvidence]
    alternatives: list[ColumnCandidate]
    warnings: list[SchemaValidationIssue]


class DatasetSchemaSummary(BaseModel):
    total_columns: int
    mapped_columns: int
    confirmed_columns: int
    unresolved_columns: int
    ambiguous_columns: int
    average_confidence: float
    primary_date_candidate: str | None
    primary_target_candidate: str | None
    revenue_candidate: str | None
    spend_candidate: str | None
    available_marketing_dimensions: list[str]
    available_performance_metrics: list[str]
    blocking_issues: list[SchemaValidationIssue]
    warnings: list[SchemaValidationIssue]
    readiness_status: str


class DatasetSchemaDetail(BaseModel):
    id: str
    dataset_id: str
    dataset_filename: str
    dataset_row_count: int
    dataset_column_count: int
    schema_version: int
    inference_version: str
    status: str
    created_at: datetime
    updated_at: datetime
    confirmed_at: datetime | None
    source_checksum: str
    sample_row_count: int
    summary: DatasetSchemaSummary
    columns: list[ColumnProfileResponse]


class SchemaInferenceRequest(BaseModel):
    force_reinfer: bool = False
    reason: str | None = Field(default=None, max_length=500)


class SchemaInferenceResponse(DatasetSchemaDetail):
    pass


class SchemaHistoryItem(BaseModel):
    id: str
    schema_version: int
    inference_version: str
    status: str
    created_at: datetime
    confirmed_at: datetime | None
    mapped_columns: int
    confirmed_columns: int
    unresolved_columns: int


class SchemaHistoryResponse(BaseModel):
    items: list[SchemaHistoryItem]


class ColumnMappingUpdateRequest(BaseModel):
    semantic_role: str
    reason: str | None = Field(default=None, max_length=500)


class ColumnMappingUpdateResponse(BaseModel):
    column: ColumnProfileResponse
    summary: DatasetSchemaSummary


class SchemaConfirmationRequest(BaseModel):
    acknowledge_warnings: bool = False


class SchemaConfirmationResponse(BaseModel):
    dataset_id: str
    schema_profile_id: str
    schema_version: int
    status: str
    confirmed_at: datetime
    summary: DatasetSchemaSummary


class SchemaStatsResponse(BaseModel):
    awaiting_review: int
    confirmed_schemas: int
    unresolved_columns: int

File: frontend/types/schema-mapping.ts
export type PhysicalType =
  | "integer"
  | "float"
  | "boolean"
  | "date"
  | "datetime"
  | "categorical"
  | "identifier"
  | "text"
  | "empty"
  | "mixed"
  | "unknown";
export type MappingStatus =
  "proposed" | "confirmed" | "manually_overridden" | "rejected" | "unresolved";
export type SchemaStatus =
  | "pending"
  | "inferred"
  | "needs_review"
  | "confirmed"
  | "superseded"
  | "failed";
export interface SemanticRoleDefinition {
  role: string;
  label: string;
  description: string;
}
export interface ColumnEvidence {
  evidence_type: string;
  description: string;
  score_contribution: number;
  observed_value: string | number | boolean | null;
  expected_pattern: string;
  severity: string;
}
export interface ColumnCandidate {
  role: string;
  confidence_score: number;
  summary_evidence: string[];
}
export interface SchemaValidationIssue {
  code: string;
  message: string;
  severity: string;
  column_names: string[];
}
export interface ColumnProfile {
  id: string;
  column_index: number;
  column_name: string;
  normalized_column_name: string;
  physical_type: PhysicalType;
  semantic_role: string;
  confidence_score: number;
  confirmation_status: MappingStatus;
  decision_source: string;
  nullable: boolean;
  null_count: number;
  sample_count: number;
  unique_count: number;
  parse_success_rate: number;
  numeric_min: number | null;
  numeric_max: number | null;
  numeric_mean: number | null;
  date_min: string | null;
  date_max: string | null;
  string_min_length: number | null;
  string_max_length: number | null;
  sample_values: string[];
  evidence: ColumnEvidence[];
  alternatives: ColumnCandidate[];
  warnings: SchemaValidationIssue[];
}
export interface DatasetSchemaSummary {
  total_columns: number;
  mapped_columns: number;
  confirmed_columns: number;
  unresolved_columns: number;
  ambiguous_columns: number;
  average_confidence: number;
  primary_date_candidate: string | null;
  primary_target_candidate: string | null;
  revenue_candidate: string | null;
  spend_candidate: string | null;
  available_marketing_dimensions: string[];
  available_performance_metrics: string[];
  blocking_issues: SchemaValidationIssue[];
  warnings: SchemaValidationIssue[];
  readiness_status: string;
}
export interface DatasetSchemaDetail {
  id: string;
  dataset_id: string;
  dataset_filename: string;
  dataset_row_count: number;
  dataset_column_count: number;
  schema_version: number;
  inference_version: string;
  status: SchemaStatus;
  created_at: string;
  updated_at: string;
  confirmed_at: string | null;
  source_checksum: string;
  sample_row_count: number;
  summary: DatasetSchemaSummary;
  columns: ColumnProfile[];
}
export interface SchemaHistoryItem {
  id: string;
  schema_version: number;
  inference_version: string;
  status: SchemaStatus;
  created_at: string;
  confirmed_at: string | null;
  mapped_columns: number;
  confirmed_columns: number;
  unresolved_columns: number;
}
export interface ColumnMappingUpdateRequest {
  semantic_role: string;
  reason?: string;
}
export interface SchemaConfirmationResponse {
  dataset_id: string;
  schema_profile_id: string;
  schema_version: number;
  status: "confirmed";
  confirmed_at: string;
  summary: DatasetSchemaSummary;
}
export interface SchemaStats {
  awaiting_review: number;
  confirmed_schemas: number;
  unresolved_columns: number;
}
